packages/contractkit/src/identity/offchain/schemas/utils.ts

```diff
@@ -0,0 +1,348 @@
+import { Err, Ok, parseJsonAsResult, Result, trimLeading0x } from '@celo/base/src'
+import { publicKeyToAddress } from '@celo/utils/lib/address'
+import { AES128Decrypt, Encrypt } from '@celo/utils/lib/ecies'
+import { EIP712Object, EIP712TypedData } from '@celo/utils/lib/sign-typed-data-utils'
+import { trimPublicKeyPrefix } from '@celo/utils/src/ecdh'
+import { createCipheriv, createHmac, randomBytes } from 'crypto'
+import { keccak256 } from 'ethereumjs-util'
+import { isLeft } from 'fp-ts/lib/Either'
+import * as t from 'io-ts'
+import { join, normalize } from 'path'
+import OffchainDataWrapper, { OffchainErrorTypes } from '../../offchain-data-wrapper'
+import {
+  InvalidDataError,
+  InvalidKey,
+  OffchainError,
+  SchemaErrors,
+  SchemaErrorTypes,
+  UnavailableKey,
+} from './errors'
+
+export interface Schema<DataType> {
+  write: (data: DataType) => Promise<SchemaErrors | void>
+  read: (from: string) => Promise<DataType>
+  readAsResult: (from: string) => Promise<Result<DataType, SchemaErrors>>
+}
+
+export interface EncryptedSchema<DataType> {
+  write: (data: DataType, to: string[], symmetricKey?: Buffer) => Promise<SchemaErrors | void>
+  read: (from: string) => Promise<DataType>
+  readAsResult: (from: string) => Promise<Result<DataType, SchemaErrors>>
+}
+
+const KEY_LENGTH = 16
+const IV_LENGTH = 16
+
+// label = PRF(ECDH(A, B), A || B || data path)
+// ciphertext path = "/cosmetic path/" || base64(label)
+function getCiphertextLabel(
+  path: string,
+  sharedSecret: Buffer,
+  senderPublicKey: string,
+  receiverPublicKey: string
+) {
+  const senderPublicKeyBuffer = Buffer.from(trimLeading0x(senderPublicKey), 'hex')
+  const receiverPublicKeyBuffer = Buffer.from(trimLeading0x(receiverPublicKey), 'hex')
+
+  const label = createHmac('blake2s256', sharedSecret)
+    .update(Buffer.concat([senderPublicKeyBuffer, receiverPublicKeyBuffer, Buffer.from(path)]))
```

Comment:
IIRC I made a comment before about normalizing the path. I see the call to `path.normalize` below, but it's possible there was some confusing. I was discussing using it on the `path` input to this function, which is used to form the label here.